File: backend/src/ai/targeting.ts
import type { BoardViewDTO, ShipKind } from "@shared/index.js";
import { FLEET_ORDER } from "../config.js";
import { idx, rowCol, expand2x2 } from "../core/coords.js";

/**
 * Targeter: derives shot priorities from an enemy BoardViewDTO. The AI calls
 * `update(view)` before each shot so density reflects the latest knowledge.
 */
export class Targeter {
  /** Last known enemy view. */
  view!: BoardViewDTO;
  /** Cells with HIT status that haven't yet been part of a SUNK ship. */
  activeHits: number[] = [];

  /** Length of every remaining (unsunk) enemy ship. */
  remainingSizes: number[] = [];

  /** Probability density per cell, recomputed in `update`. */
  density: number[] = [];

  constructor(view: BoardViewDTO) {
    this.update(view);
  }

  update(view: BoardViewDTO): void {
    this.view = view;
    this.activeHits = [];
    for (let i = 0; i < view.cells.length; i++) {
      if (view.cells[i] === "HIT") this.activeHits.push(i);
    }
    // Remaining ship sizes = full roster minus those whose kind has been sunk.
    const sunkKinds = new Set<ShipKind>(
      view.ships.filter((s) => s.sunk).map((s) => s.kind),
    );
    const all = lengthForEach();
    const remaining: number[] = [];
    for (const k of FLEET_ORDER) {
      if (!sunkKinds.has(k)) remaining.push(all[k]);
    }
    this.remainingSizes = remaining;
    this.density = this.computeDensity();
  }

  /** Single-cell densest unattacked cell. */
  bestSingleTarget(): number {
    const view = this.view;
    let best = -1;
    let bestScore = -1;
    for (let i = 0; i < view.cells.length; i++) {
      if (view.cells[i] !== "UNKNOWN") continue;
      const s = this.density[i] ?? 0;
      if (s > bestScore) {
        bestScore = s;
        best = i;
      }
    }
    // Fallback: any unattacked cell.
    if (best < 0) {
      for (let i = 0; i < view.cells.length; i++) {
        if (view.cells[i] === "UNKNOWN") return i;
      }
    }
    return best;
  }

  /**
   * Best 2x2 anchor by sum-of-density over its 4 cells. Cells outside the
   * board are penalized; cells that are already attacked contribute 0.
   */
  bestAreaTarget(): { anchor: number; score: number } {
    const size = this.view.size;
    let bestAnchor = -1;
    let bestScore = -Infinity;
    for (let r = 0; r < size - 1; r++) {
      for (let c = 0; c < size - 1; c++) {
        const anchor = idx(r, c, size);
        const cells = expand2x2(anchor, size);
        let score = 0;
        let unattacked = 0;
        for (const cell of cells) {
          if (this.view.cells[cell] === "UNKNOWN") {
            score += this.density[cell] ?? 0;
            unattacked += 1;
          }
        }
        // Require at least 2 unattacked cells to make the 2x2 worth firing.
        if (unattacked < 2) continue;
        if (score > bestScore) {
          bestScore = score;
          bestAnchor = anchor;
        }
      }
    }
    if (bestAnchor < 0) {
      // Fallback: lowest legal anchor with at least 1 unattacked cell.
      for (let r = 0; r < size - 1; r++) {
        for (let c = 0; c < size - 1; c++) {
          const anchor = idx(r, c, size);
          const cells = expand2x2(anchor, size);
          if (cells.some((cc) => this.view.cells[cc] === "UNKNOWN")) {
            return { anchor, score: 0 };
          }
        }
      }
      // Truly nothing left. Default to (0,0).
      return { anchor: 0, score: 0 };
    }
    return { anchor: bestAnchor, score: bestScore };
  }

  private computeDensity(): number[] {
    const size = this.view.size;
    const density = new Array<number>(size * size).fill(0);
    const targetMode = this.activeHits.length > 0;
    const hitSet = new Set(this.activeHits);

    const blocked = (cell: number) =>
      this.view.cells[cell] === "MISS" || this.view.cells[cell] === "SUNK";

    for (const shipSize of this.remainingSizes) {
      // Horizontal placements.
      for (let r = 0; r < size; r++) {
        for (let c = 0; c <= size - shipSize; c++) {
          let ok = true;
          let overlapsHit = false;
          for (let k = 0; k < shipSize; k++) {
            const cell = idx(r, c + k, size);
            if (blocked(cell)) {
              ok = false;
              break;
            }
            if (hitSet.has(cell)) overlapsHit = true;
          }
          if (!ok) continue;
          if (targetMode && !overlapsHit) continue;
          for (let k = 0; k < shipSize; k++) {
            const cell = idx(r, c + k, size);
            if (this.view.cells[cell] !== "UNKNOWN") continue;
            density[cell] = (density[cell] ?? 0) + 1;
          }
        }
      }
      // Vertical placements.
      for (let c = 0; c < size; c++) {
        for (let r = 0; r <= size - shipSize; r++) {
          let ok = true;
          let overlapsHit = false;
          for (let k = 0; k < shipSize; k++) {
            const cell = idx(r + k, c, size);
            if (blocked(cell)) {
              ok = false;
              break;
            }
            if (hitSet.has(cell)) overlapsHit = true;
          }
          if (!ok) continue;
          if (targetMode && !overlapsHit) continue;
          for (let k = 0; k < shipSize; k++) {
            const cell = idx(r + k, c, size);
            if (this.view.cells[cell] !== "UNKNOWN") continue;
            density[cell] = (density[cell] ?? 0) + 1;
          }
        }
      }
    }
    return density;
  }
}

// Static lookup: ShipKind -> default length. We avoid importing the Ship
// classes here to keep this module dependency-light.
function lengthForEach(): Record<ShipKind, number> {
  return {
    AIRCRAFT_CARRIER: 8,
    BATTLESHIP: 6,
    CRUISER: 4,
    FRIGATE: 3,
    SUBMARINE: 2,
  };
}

/** Get the row-col of the center of mass of MISS cells (for sub evasion). */
export function missCenter(view: BoardViewDTO): { row: number; col: number } | null {
  let sumR = 0;
  let sumC = 0;
  let n = 0;
  for (let i = 0; i < view.cells.length; i++) {
    if (view.cells[i] === "MISS" || view.cells[i] === "HIT") {
      const { row, col } = rowCol(i, view.size);
      sumR += row;
      sumC += col;
      n += 1;
    }
  }
  if (n === 0) return null;
  return { row: sumR / n, col: sumC / n };
}

File: backend-rs/src/ai/targeting.rs
use std::collections::HashSet;

use crate::config::FLEET_ORDER;
use crate::core::coords::{expand_2x2, idx, row_col};
use crate::shared::{BoardViewDTO, CellKnowledge, ShipKind};

/// Targeter: derives shot priorities from an enemy BoardViewDTO.
pub struct Targeter {
    pub view: BoardViewDTO,
    pub active_hits: Vec<u32>,
    pub remaining_sizes: Vec<u32>,
    pub density: Vec<i32>,
}

impl Targeter {
    pub fn new(view: BoardViewDTO) -> Self {
        let mut t = Self {
            view,
            active_hits: Vec::new(),
            remaining_sizes: Vec::new(),
            density: Vec::new(),
        };
        t.update();
        t
    }

    pub fn update(&mut self) {
        self.active_hits.clear();
        for (i, c) in self.view.cells.iter().enumerate() {
            if matches!(c, CellKnowledge::Hit) {
                self.active_hits.push(i as u32);
            }
        }
        let sunk_kinds: HashSet<ShipKind> = self
            .view
            .ships
            .iter()
            .filter(|s| s.sunk)
            .map(|s| s.kind)
            .collect();
        let lengths = length_for_each();
        let mut remaining = Vec::new();
        for k in FLEET_ORDER.iter().copied() {
            if !sunk_kinds.contains(&k) {
                remaining.push(lengths(k));
            }
        }
        self.remaining_sizes = remaining;
        self.density = self.compute_density();
    }

    /// Single-cell densest unattacked cell. Returns -1 if none.
    pub fn best_single_target(&self) -> i64 {
        let view = &self.view;
        let mut best: i64 = -1;
        let mut best_score: i32 = -1;
        for i in 0..view.cells.len() {
            if !matches!(view.cells[i], CellKnowledge::Unknown) {
                continue;
            }
            let s = self.density.get(i).copied().unwrap_or(0);
            if s > best_score {
                best_score = s;
                best = i as i64;
            }
        }
        if best < 0 {
            for i in 0..view.cells.len() {
                if matches!(view.cells[i], CellKnowledge::Unknown) {
                    return i as i64;
                }
            }
        }
        best
    }

    /// Best 2x2 anchor by sum-of-density. Returns `(anchor, score)`.
    pub fn best_area_target(&self) -> (u32, i32) {
        let size = self.view.size;
        let mut best_anchor: i64 = -1;
        let mut best_score: i32 = i32::MIN;
        for r in 0..((size as i32) - 1) {
            for c in 0..((size as i32) - 1) {
                let anchor = idx(r, c, size);
                let cells = expand_2x2(anchor, size);
                let mut score = 0i32;
                let mut unattacked = 0i32;
                for cell in &cells {
                    if matches!(self.view.cells[*cell as usize], CellKnowledge::Unknown) {
                        score += self.density.get(*cell as usize).copied().unwrap_or(0);
                        unattacked += 1;
                    }
                }
                if unattacked < 2 {
                    continue;
                }
                if score > best_score {
                    best_score = score;
                    best_anchor = anchor as i64;
                }
            }
        }
        if best_anchor < 0 {
            for r in 0..((size as i32) - 1) {
                for c in 0..((size as i32) - 1) {
                    let anchor = idx(r, c, size);
                    let cells = expand_2x2(anchor, size);
                    if cells
                        .iter()
                        .any(|cc| matches!(self.view.cells[*cc as usize], CellKnowledge::Unknown))
                    {
                        return (anchor, 0);
                    }
                }
            }
            return (0, 0);
        }
        (best_anchor as u32, best_score)
    }

    fn compute_density(&self) -> Vec<i32> {
        let size = self.view.size;
        let n = (size * size) as usize;
        let mut density = vec![0i32; n];
        let target_mode = !self.active_hits.is_empty();
        let hit_set: HashSet<u32> = self.active_hits.iter().copied().collect();

        let blocked = |cell: u32| {
            matches!(
                self.view.cells[cell as usize],
                CellKnowledge::Miss | CellKnowledge::Sunk
            )
        };

        for &ship_size in &self.remaining_sizes {
            // Horizontal placements.
            for r in 0..(size as i32) {
                for c in 0..=((size as i32) - (ship_size as i32)) {
                    let mut ok = true;
                    let mut overlaps_hit = false;
                    for k in 0..(ship_size as i32) {
                        let cell = idx(r, c + k, size);
                        if blocked(cell) {
                            ok = false;
                            break;
                        }
                        if hit_set.contains(&cell) {
                            overlaps_hit = true;
                        }
                    }
                    if !ok {
                        continue;
                    }
                    if target_mode && !overlaps_hit {
                        continue;
                    }
                    for k in 0..(ship_size as i32) {
                        let cell = idx(r, c + k, size);
                        if !matches!(self.view.cells[cell as usize], CellKnowledge::Unknown) {
                            continue;
                        }
                        density[cell as usize] += 1;
                    }
                }
            }
            // Vertical placements.
            for c in 0..(size as i32) {
                for r in 0..=((size as i32) - (ship_size as i32)) {
                    let mut ok = true;
                    let mut overlaps_hit = false;
                    for k in 0..(ship_size as i32) {
                        let cell = idx(r + k, c, size);
                        if blocked(cell) {
                            ok = false;
                            break;
                        }
                        if hit_set.contains(&cell) {
                            overlaps_hit = true;
                        }
                    }
                    if !ok {
                        continue;
                    }
                    if target_mode && !overlaps_hit {
                        continue;
                    }
                    for k in 0..(ship_size as i32) {
                        let cell = idx(r + k, c, size);
                        if !matches!(self.view.cells[cell as usize], CellKnowledge::Unknown) {
                            continue;
                        }
                        density[cell as usize] += 1;
                    }
                }
            }
        }
        density
    }
}

fn length_for_each() -> impl Fn(ShipKind) -> u32 {
    |k| match k {
        ShipKind::AircraftCarrier => 8,
        ShipKind::Battleship => 6,
        ShipKind::Cruiser => 4,
        ShipKind::Frigate => 3,
        ShipKind::Submarine => 2,
    }
}

/// Center of mass of MISS/HIT cells (for sub evasion). Returns None if none.
pub fn miss_center(view: &BoardViewDTO) -> Option<(f64, f64)> {
    let mut sum_r = 0.0;
    let mut sum_c = 0.0;
    let mut n = 0.0;
    for i in 0..view.cells.len() {
        if matches!(view.cells[i], CellKnowledge::Miss | CellKnowledge::Hit) {
            let (row, col) = row_col(i as u32, view.size);
            sum_r += row as f64;
            sum_c += col as f64;
            n += 1.0;
        }
    }
    if n == 0.0 {
        return None;
    }
    Some((sum_r / n, sum_c / n))
}
